rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java

```diff
@@ -691,39 +692,54 @@ private FederationChangeResponseCode commitPendingFederationAccordingToActivatio
      * The federation change info is preserved, and the commitment with the voted federation is logged.
      */
     private FederationChangeResponseCode legacyCommitPendingFederation(PendingFederation currentPendingFederation, BridgeEventLogger eventLogger) {
-        moveUTXOsFromNewToOldFederation();
-
-        // set old and new federations
-        Federation activeFederation = getActiveFederation();
-        provider.setOldFederation(activeFederation);
         Federation newFederation = buildFederationFromPendingFederation(currentPendingFederation);
-        provider.setNewFederation(newFederation);
+        handoverToNewFederation(newFederation);
 
         clearPendingFederationVoting();
 
-        if (activations.isActive(RSKIP186)) {
-            // since we are creating the to-be-active-fed in this block,
-            // its creation block height is this block number
-            saveFederationChangeInfo(rskExecutionBlock.getNumber());
-        }
-
         Federation currentOldFederation = provider.getOldFederation(constants, activations);
         Federation currentNewFederation = provider.getNewFederation(constants, activations);
         logCommitmentWithVotedFederation(eventLogger, currentOldFederation, currentNewFederation);
 
         return FederationChangeResponseCode.SUCCESSFUL;
     }
 
+    public void commitProposedFederation() {
+        Federation proposedFederation = provider.getProposedFederation(constants, activations)
+                .orElseThrow(IllegalStateException::new);
+
+        handoverToNewFederation(proposedFederation);
+        clearProposedFederation();
+    }
+
+    private void handoverToNewFederation(Federation newFederation) {
+        moveUTXOsFromNewToOldFederation();
+
+        setOldAndNewFederations(getActiveFederation(), newFederation);
+
+        if (activations.isActive(RSKIP186)) {
+            saveLastRetiredFederationScript();
+            provider.setNextFederationCreationBlockHeight(newFederation.getCreationBlockNumber());
+        }
+    }
+
+    private void setOldAndNewFederations(Federation oldFederation, Federation newFederation) {
+        provider.setOldFederation(oldFederation);
+        provider.setNewFederation(newFederation);
+    }
+
     private void moveUTXOsFromNewToOldFederation() {
-        List<UTXO> utxosToMove = new ArrayList<>(provider.getNewFederationBtcUTXOs(constants.getBtcParams(), activations));
+        // since the current active fed reference will change from being 'new' to 'old',
+        // we have to change the UTXOs reference to match it
+        List<UTXO> activeFederationUTXOs = new ArrayList<>(provider.getNewFederationBtcUTXOs(constants.getBtcParams(), activations));
```

Comment:
why pass the list to the constructor of another list?

Reply:
I think is to create a shallow copy of the list before clear the original one. Btw, you can do this instead:

```suggestion
List<UTXO> activeFederationUTXOs = provider.getNewFederationBtcUTXOs(constants.getBtcParams(), activations).stream().toList();
```

Reply:
Or simply use List.copyOf